File: cairo.cpp
#include <cairo/cairo.h>
#include <SWI-cpp2.h>
#include <string>
#include "ptr.h"

// ---------------------------------------------------------------------------
// Cairo has two main handle types exposed here:
//
//   cairo_surface_t  -- a drawing target (here always an image surface, i.e.
//                       a CPU-side pixel buffer).
//   cairo_t          -- a drawing context.  It holds all graphics state
//                       (source colour, line width, antialias mode, the
//                       current path, ...) and is bound to a target surface
//                       at creation time.  cairo_create() takes its own
//                       reference on the surface, but we additionally keep a
//                       registered PlAtom reference to the surface blob so
//                       that Prolog GC can never collect the surface blob
//                       out from under a live context blob (Prolog GC order
//                       is not guaranteed).  This mirrors the
//                       SDLRendererBlob->window and SDLTextureBlob->renderer
//                       parent links in sdl.cpp.
//
// Each handle is wrapped in a PlBlob subclass with RAII semantics: the
// destructor / destroy() calls the matching cairo_*_destroy() and releases
// the parent reference.  Foreign predicates are named with a trailing
// underscore (e.g. cairo_create_/2); the Prolog side in prolog/cairo.pl
// provides the public name without the underscore after type-checking.
// ---------------------------------------------------------------------------

struct CairoSurfaceBlob;

static PL_blob_t cairo_surface_blob =
    PL_BLOB_DEFINITION(CairoSurfaceBlob, "cairo_surface_blob");

struct CairoSurfaceBlob : public PlBlob {
  cairo_surface_t *surface_;
  PlAtom parent_; // keeps a PtrBlob alive (for for_data surfaces)

  explicit CairoSurfaceBlob() : PlBlob(&cairo_surface_blob), parent_(PlAtom::null) {}

  explicit CairoSurfaceBlob(cairo_surface_t *surface)
      : PlBlob(&cairo_surface_blob), surface_(surface), parent_(PlAtom::null) {}

  explicit CairoSurfaceBlob(cairo_surface_t *surface, PlAtom parent)
      : PlBlob(&cairo_surface_blob), surface_(surface), parent_(parent) {
    parent_.register_ref();
  }

  PL_BLOB_SIZE

  void portray(PlStream &strm) const {
    strm.printf("cairo_surface_blob<%p>(%p)", this, surface_);
  }

  void destroy() noexcept {
    if (surface_ != NULL) {
      cairo_surface_destroy(surface_);
      surface_ = NULL;
    }
    if (parent_.not_null()) {
      parent_.unregister_ref();
      parent_.set_null();
    }
  }

  virtual ~CairoSurfaceBlob() noexcept { destroy(); }
};

PREDICATE(cairo_surface_blob_portray, 2) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A2, cairo_surface_blob);
  PlStream strm(A1, 0);
  ref->portray(strm);
  return true;
}

struct CairoContextBlob;

static PL_blob_t cairo_context_blob =
    PL_BLOB_DEFINITION(CairoContextBlob, "cairo_context_blob");

struct CairoContextBlob : public PlBlob {
  cairo_t *cr_;
  PlAtom parent_; // keeps the owning surface alive while the ctx lives

  explicit CairoContextBlob()
      : PlBlob(&cairo_context_blob), cr_(NULL), parent_(PlAtom::null) {}

  explicit CairoContextBlob(cairo_t *cr, PlAtom surface)
      : PlBlob(&cairo_context_blob), cr_(cr), parent_(surface) {
    parent_.register_ref();
  }

  PL_BLOB_SIZE

  void portray(PlStream &strm) const {
    strm.printf("cairo_context_blob<%p>(%p)", this, cr_);
  }

  void destroy() noexcept {
    if (cr_ != NULL) {
      cairo_destroy(cr_);
      cr_ = NULL;
    }
    if (parent_.not_null()) {
      parent_.unregister_ref();
      parent_.set_null();
    }
  }

  virtual ~CairoContextBlob() noexcept { destroy(); }
};

PREDICATE(cairo_context_blob_portray, 2) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A2, cairo_context_blob);
  PlStream strm(A1, 0);
  ref->portray(strm);
  return true;
}

// ---------------------------------------------------------------------------
// Image surfaces: rendering to memory buffers.
// ---------------------------------------------------------------------------

PREDICATE(cairo_image_surface_create_, 4) {
  cairo_format_t format = (cairo_format_t)A2.as_int();
  cairo_surface_t *surface =
      cairo_image_surface_create(format, A3.as_int(), A4.as_int());
  if (cairo_surface_status(surface) != CAIRO_STATUS_SUCCESS) {
    cairo_surface_destroy(surface);
    throw PlUnknownError(cairo_status_to_string(
        cairo_surface_status(surface)));
  }
  auto ref =
      std::unique_ptr<PlBlob>(new CairoSurfaceBlob(surface));
  return A1.unify_blob(&ref);
}

PREDICATE(cairo_image_surface_create_for_data_, 6) {
  auto ptr_ref = PlBlobV<PtrBlob>::cast_ex(A2, *ptr_blob_type());
  cairo_format_t format = (cairo_format_t)A3.as_int();
  cairo_surface_t *surface = cairo_image_surface_create_for_data(
      (unsigned char *)ptr_ref->ptr, format, A4.as_int(), A5.as_int(), A6.as_int());
  if (cairo_surface_status(surface) != CAIRO_STATUS_SUCCESS) {
    cairo_surface_destroy(surface);
    throw PlUnknownError(cairo_status_to_string(
        cairo_surface_status(surface)));
  }
  // Parent = PtrBlob → keeps the pixel buffer alive (owned by SDL texture).
  // cairo does NOT free the pixel data on surface_destroy.
  auto ref = std::unique_ptr<PlBlob>(
      new CairoSurfaceBlob(surface, ptr_ref->symbol_));
  return A1.unify_blob(&ref);
}

PREDICATE(cairo_surface_destroy_, 1) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  ref->destroy();
  return true;
}

PREDICATE(cairo_surface_flush_, 1) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  cairo_surface_flush(ref->surface_);
  return true;
}

PREDICATE(cairo_image_surface_get_data_, 2) {
  auto surface_ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  unsigned char *data = cairo_image_surface_get_data(surface_ref->surface_);
  // Return a PtrBlob that holds a parent_ ref to the surface, keeping
  // the pixel buffer alive while the pointer view is referenced.  The
  // pointer itself never escapes to Prolog as a raw integer.
  auto ptr = std::unique_ptr<PlBlob>(
      new PtrBlob(data, surface_ref->symbol_));
  return A2.unify_blob(&ptr);
}

PREDICATE(cairo_image_surface_get_width_, 2) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  return A2.unify_integer(cairo_image_surface_get_width(ref->surface_));
}

PREDICATE(cairo_image_surface_get_height_, 2) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  return A2.unify_integer(cairo_image_surface_get_height(ref->surface_));
}

PREDICATE(cairo_image_surface_get_stride_, 2) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  return A2.unify_integer(cairo_image_surface_get_stride(ref->surface_));
}

PREDICATE(cairo_image_surface_get_format_, 2) {
  auto ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A1, cairo_surface_blob);
  return A2.unify_integer(
      (int)cairo_image_surface_get_format(ref->surface_));
}

// ---------------------------------------------------------------------------
// Context: create / destroy / status.
// ---------------------------------------------------------------------------

PREDICATE(cairo_create_, 2) {
  auto surface_ref = PlBlobV<CairoSurfaceBlob>::cast_ex(A2, cairo_surface_blob);
  cairo_t *cr = cairo_create(surface_ref->surface_);
  if (cairo_status(cr) != CAIRO_STATUS_SUCCESS) {
    cairo_destroy(cr);
    throw PlUnknownError(cairo_status_to_string(cairo_status(cr)));
  }
  auto ref = std::unique_ptr<PlBlob>(
      new CairoContextBlob(cr, surface_ref->symbol_));
  return A1.unify_blob(&ref);
}

PREDICATE(cairo_destroy_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  ref->destroy();
  return true;
}

// ---------------------------------------------------------------------------
// Source colour.
// ---------------------------------------------------------------------------

PREDICATE(cairo_set_source_rgba_, 5) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_set_source_rgba(ref->cr_, A2.as_double(), A3.as_double(),
                        A4.as_double(), A5.as_double());
  return true;
}

// ---------------------------------------------------------------------------
// Graphics state.
// ---------------------------------------------------------------------------

PREDICATE(cairo_set_antialias_, 2) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_set_antialias(ref->cr_, (cairo_antialias_t)A2.as_int());
  return true;
}

PREDICATE(cairo_set_line_width_, 2) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_set_line_width(ref->cr_, A2.as_double());
  return true;
}

PREDICATE(cairo_set_line_cap_, 2) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_set_line_cap(ref->cr_, (cairo_line_cap_t)A2.as_int());
  return true;
}

PREDICATE(cairo_set_line_join_, 2) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_set_line_join(ref->cr_, (cairo_line_join_t)A2.as_int());
  return true;
}

// ---------------------------------------------------------------------------
// Paths.
// ---------------------------------------------------------------------------

PREDICATE(cairo_new_path_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_new_path(ref->cr_);
  return true;
}

PREDICATE(cairo_new_sub_path_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_new_sub_path(ref->cr_);
  return true;
}

PREDICATE(cairo_close_path_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_close_path(ref->cr_);
  return true;
}

PREDICATE(cairo_move_to_, 3) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_move_to(ref->cr_, A2.as_double(), A3.as_double());
  return true;
}

PREDICATE(cairo_line_to_, 3) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_line_to(ref->cr_, A2.as_double(), A3.as_double());
  return true;
}

PREDICATE(cairo_rectangle_, 5) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_rectangle(ref->cr_, A2.as_double(), A3.as_double(),
                  A4.as_double(), A5.as_double());
  return true;
}

PREDICATE(cairo_arc_, 6) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_arc(ref->cr_, A2.as_double(), A3.as_double(), A4.as_double(),
            A5.as_double(), A6.as_double());
  return true;
}

PREDICATE(cairo_arc_negative_, 6) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_arc_negative(ref->cr_, A2.as_double(), A3.as_double(),
                     A4.as_double(), A5.as_double(), A6.as_double());
  return true;
}

PREDICATE(cairo_curve_to_, 7) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_curve_to(ref->cr_, A2.as_double(), A3.as_double(), A4.as_double(),
                 A5.as_double(), A6.as_double(), A7.as_double());
  return true;
}

// ---------------------------------------------------------------------------
// Drawing: fill / stroke / paint.  These consume the current path (except
// the *_preserve variants).
// ---------------------------------------------------------------------------

PREDICATE(cairo_fill_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_fill(ref->cr_);
  return true;
}

PREDICATE(cairo_fill_preserve_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_fill_preserve(ref->cr_);
  return true;
}

PREDICATE(cairo_stroke_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_stroke(ref->cr_);
  return true;
}

PREDICATE(cairo_stroke_preserve_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_stroke_preserve(ref->cr_);
  return true;
}

PREDICATE(cairo_paint_, 1) {
  auto ref = PlBlobV<CairoContextBlob>::cast_ex(A1, cairo_context_blob);
  cairo_paint(ref->cr_);
  return true;
}

File: prolog/cairo.pl
:- module(cairo, [cairo_image_surface_create/4,
                  cairo_image_surface_create_for_data/6,
                  cairo_surface_destroy/1, cairo_surface_flush/1,
                  cairo_image_surface_get_data/2, cairo_image_surface_get_width/2,
                  cairo_image_surface_get_height/2, cairo_image_surface_get_stride/2,
                  cairo_image_surface_get_format/2,
                  cairo_create/2, cairo_destroy/1,
                  cairo_set_source_rgba/5,
                  cairo_set_antialias/2, cairo_set_line_width/2,
                  cairo_set_line_cap/2, cairo_set_line_join/2,
                  cairo_new_path/1, cairo_new_sub_path/1, cairo_close_path/1,
                  cairo_move_to/3, cairo_line_to/3, cairo_rectangle/5,
                  cairo_arc/6, cairo_arc_negative/6, cairo_curve_to/7,
                  cairo_fill/1, cairo_fill_preserve/1,
                  cairo_stroke/1, cairo_stroke_preserve/1, cairo_paint/1,
                  cairo_format/2, cairo_antialias/2, cairo_line_cap/2,
                  cairo_line_join/2
                  ]).
:- use_foreign_library(foreign(cairo)).

:- multifile error:has_type/2.
error:has_type(cairo_surface_blob, X) :- blob(X, cairo_surface_blob).
error:has_type(cairo_context_blob, X) :- blob(X, cairo_context_blob).
error:has_type(cairo_format,     X) :- cairo_format(X, _).
error:has_type(cairo_antialias,  X) :- cairo_antialias(X, _).
error:has_type(cairo_line_cap,   X) :- cairo_line_cap(X, _).
error:has_type(cairo_line_join,  X) :- cairo_line_join(X, _).

% Rich error messages listing the valid atoms for each enum.
:- multifile prolog:error_message//1.
prolog:error_message(type_error(cairo_format, Culprit)) -->
   { findall(F, cairo_format(F, _), Fs) },
   [ 'cairo_format (one of ~q), found ~q'-[Fs, Culprit] ].
prolog:error_message(type_error(cairo_antialias, Culprit)) -->
   { findall(F, cairo_antialias(F, _), Fs) },
   [ 'cairo_antialias (one of ~q), found ~q'-[Fs, Culprit] ].
prolog:error_message(type_error(cairo_line_cap, Culprit)) -->
   { findall(F, cairo_line_cap(F, _), Fs) },
   [ 'cairo_line_cap (one of ~q), found ~q'-[Fs, Culprit] ].
prolog:error_message(type_error(cairo_line_join, Culprit)) -->
   { findall(F, cairo_line_join(F, _), Fs) },
   [ 'cairo_line_join (one of ~q), found ~q'-[Fs, Culprit] ].

user:portray(Surface) :-
   blob(Surface, cairo_surface_blob), !,
   cairo_surface_blob_portray(current_output, Surface).
user:portray(Context) :-
   blob(Context, cairo_context_blob), !,
   cairo_context_blob_portray(current_output, Context).

% --- enum tables ------------------------------------------------------------
% cairo_format_t: see cairo/cairo.h.  argb32 is the natural pairing with
% SDL_PIXELFORMAT_ARGB32 (native-endian 0xAARRGGBB).
cairo_format(invalid, -1).
cairo_format(argb32, 0).
cairo_format(rgb24, 1).
cairo_format(a8, 2).
cairo_format(a1, 3).
cairo_format(rgb16_565, 4).
cairo_format(rgb30, 5).

% cairo_antialias_t
cairo_antialias(default, 0).
cairo_antialias(none, 1).
cairo_antialias(gray, 2).
cairo_antialias(subpixel, 3).
cairo_antialias(fast, 4).
cairo_antialias(good, 5).
cairo_antialias(best, 6).

% cairo_line_cap_t
cairo_line_cap(butt, 0).
cairo_line_cap(round, 1).
cairo_line_cap(square, 2).

% cairo_line_join_t
cairo_line_join(miter, 0).
cairo_line_join(round, 1).
cairo_line_join(bevel, 2).

% --- image surfaces ---------------------------------------------------------

cairo_image_surface_create(Surface, Format, Width, Height) :-
   must_be(var, Surface),
   must_be(cairo_format, Format),
   must_be(positive_integer, Width),
   must_be(positive_integer, Height),
   cairo_format(Format, IntFormat),
   cairo_image_surface_create_(Surface, IntFormat, Width, Height).

% Creates a cairo image surface backed by external pixel data (e.g. an
% SDL locked texture).  The Pixels PtrBlob must stay alive while the
% surface is in use — the surface blob holds a parent ref to it.
cairo_image_surface_create_for_data(Surface, Pixels, Format, Width, Height, Stride) :-
   must_be(var, Surface),
   must_be(ptr_blob, Pixels),
   must_be(cairo_format, Format),
   must_be(positive_integer, Width),
   must_be(positive_integer, Height),
   must_be(integer, Stride),
   cairo_format(Format, IntFormat),
   cairo_image_surface_create_for_data_(Surface, Pixels, IntFormat, Width, Height, Stride).

cairo_surface_destroy(Surface) :-
   must_be(cairo_surface_blob, Surface),
   cairo_surface_destroy_(Surface).

cairo_surface_flush(Surface) :-
   must_be(cairo_surface_blob, Surface),
   cairo_surface_flush_(Surface).

cairo_image_surface_get_data(Surface, Data) :-
   must_be(cairo_surface_blob, Surface),
   must_be(var, Data),
   cairo_image_surface_get_data_(Surface, Data).

cairo_image_surface_get_width(Surface, Width) :-
   must_be(cairo_surface_blob, Surface),
   must_be(var, Width),
   cairo_image_surface_get_width_(Surface, Width).

cairo_image_surface_get_height(Surface, Height) :-
   must_be(cairo_surface_blob, Surface),
   must_be(var, Height),
   cairo_image_surface_get_height_(Surface, Height).

cairo_image_surface_get_stride(Surface, Stride) :-
   must_be(cairo_surface_blob, Surface),
   must_be(var, Stride),
   cairo_image_surface_get_stride_(Surface, Stride).

cairo_image_surface_get_format(Surface, Format) :-
   must_be(cairo_surface_blob, Surface),
   must_be(var, Format),
   cairo_image_surface_get_format_(Surface, IntFormat),
   cairo_format(Format, IntFormat).

% --- context ----------------------------------------------------------------

cairo_create(Context, Surface) :-
   must_be(var, Context),
   must_be(cairo_surface_blob, Surface),
   cairo_create_(Context, Surface).

cairo_destroy(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_destroy_(Context).

% --- source colour ----------------------------------------------------------
% Colour components are doubles in 0.0..1.0 (unlike SDL's 0..255 ints).

cairo_set_source_rgba(Context, R, G, B, A) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [R, G, B, A]),
   cairo_set_source_rgba_(Context, R, G, B, A).

% --- graphics state ---------------------------------------------------------

cairo_set_antialias(Context, Antialias) :-
   must_be(cairo_context_blob, Context),
   must_be(cairo_antialias, Antialias),
   cairo_antialias(Antialias, Int),
   cairo_set_antialias_(Context, Int).

cairo_set_line_width(Context, Width) :-
   must_be(cairo_context_blob, Context),
   must_be(number, Width),
   cairo_set_line_width_(Context, Width).

cairo_set_line_cap(Context, Cap) :-
   must_be(cairo_context_blob, Context),
   must_be(cairo_line_cap, Cap),
   cairo_line_cap(Cap, Int),
   cairo_set_line_cap_(Context, Int).

cairo_set_line_join(Context, Join) :-
   must_be(cairo_context_blob, Context),
   must_be(cairo_line_join, Join),
   cairo_line_join(Join, Int),
   cairo_set_line_join_(Context, Int).

% --- paths ------------------------------------------------------------------

cairo_new_path(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_new_path_(Context).

cairo_new_sub_path(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_new_sub_path_(Context).

cairo_close_path(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_close_path_(Context).

cairo_move_to(Context, X, Y) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [X, Y]),
   cairo_move_to_(Context, X, Y).

cairo_line_to(Context, X, Y) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [X, Y]),
   cairo_line_to_(Context, X, Y).

cairo_rectangle(Context, X, Y, Width, Height) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [X, Y, Width, Height]),
   cairo_rectangle_(Context, X, Y, Width, Height).

cairo_arc(Context, Xc, Yc, Radius, Angle1, Angle2) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [Xc, Yc, Radius, Angle1, Angle2]),
   cairo_arc_(Context, Xc, Yc, Radius, Angle1, Angle2).

cairo_arc_negative(Context, Xc, Yc, Radius, Angle1, Angle2) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [Xc, Yc, Radius, Angle1, Angle2]),
   cairo_arc_negative_(Context, Xc, Yc, Radius, Angle1, Angle2).

cairo_curve_to(Context, X1, Y1, X2, Y2, X3, Y3) :-
   must_be(cairo_context_blob, Context),
   maplist(must_be(number), [X1, Y1, X2, Y2, X3, Y3]),
   cairo_curve_to_(Context, X1, Y1, X2, Y2, X3, Y3).

% --- drawing ----------------------------------------------------------------

cairo_fill(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_fill_(Context).

cairo_fill_preserve(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_fill_preserve_(Context).

cairo_stroke(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_stroke_(Context).

cairo_stroke_preserve(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_stroke_preserve_(Context).

cairo_paint(Context) :-
   must_be(cairo_context_blob, Context),
   cairo_paint_(Context).
